src/app/CommandSender.cpp

```diff
@@ -62,24 +62,47 @@ CHIP_ERROR CommandSender::AllocateBuffer()
 
 CHIP_ERROR CommandSender::SendCommandRequest(const SessionHandle & session, System::Clock::Timeout timeout)
 {
-    VerifyOrReturnError(mState == State::AddedCommand, CHIP_ERROR_INCORRECT_STATE);
+    CHIP_ERROR err      = CHIP_NO_ERROR;
+    bool isGroupCommand = false;
+    VerifyOrExit(mState == State::AddedCommand, err = CHIP_ERROR_INCORRECT_STATE);
 
-    ReturnErrorOnFailure(Finalize(mPendingInvokeData));
+    err = Finalize(mPendingInvokeData);
+    SuccessOrExit(err);
 
     // Create a new exchange context.
     mpExchangeCtx = mpExchangeMgr->NewContext(session, this);
-    VerifyOrReturnError(mpExchangeCtx != nullptr, CHIP_ERROR_NO_MEMORY);
+    VerifyOrExit(mpExchangeCtx != nullptr, err = CHIP_ERROR_NO_MEMORY);
 
+    isGroupCommand = mpExchangeCtx->IsGroupExchangeContext();
     mpExchangeCtx->SetResponseTimeout(timeout);
 
     if (mTimedInvokeTimeoutMs.HasValue())
     {
-        ReturnErrorOnFailure(TimedRequest::Send(mpExchangeCtx, mTimedInvokeTimeoutMs.Value()));
+        err = TimedRequest::Send(mpExchangeCtx, mTimedInvokeTimeoutMs.Value());
+        SuccessOrExit(err);
         MoveToState(State::AwaitingTimedStatus);
-        return CHIP_NO_ERROR;
     }
+    else
+    {
+        err = SendInvokeRequest();
+    }
+    SuccessOrExit(err);
 
-    return SendInvokeRequest();
+exit:
+    if (mpCallback != nullptr)
+    {
+        if (err != CHIP_NO_ERROR)
+        {
+            mpCallback->OnError(this, StatusIB(Protocols::InteractionModel::Status::Failure), err);
+        }
+    }
+
+    if (isGroupCommand)
+    {
+        Close();
```

Comment:
This has two problems:
1.  The call to Close() is happening even in error conditions.  This causes OnDone to be called in addition to an error return, which will lead to double-delete.
2. The call to OnDone() happens synchronously under SendCommandRequest.  That's not what the API documentation says, and is not what consumers expect.

What should happen here is that the Close should happen only if err is success, and should be queued to happen async.

Reply:
Agreed on 1).

As for 2), I don't see why we cannot amend the expectation to state that on success and for a group command, the OnDone is called synchronously. I'd very much like to avoid having to do bottom-half/post calls in here if we could.

Reply:
We could amend the expectation, but it's a bit harder for the caller to manage: now they may have to keep track of whether their objects already got destroyed or not when this call returns and branch accordingly.

In general, callbacks that sometimes happen sync and sometimes happen async are very error-prone to work with and tend to lead to a lot of bugs.  Hence the suggestion to just stick to async...

Reply:
I think, in the case of group communication, `onDone` should be called in sync and the documentation updated.
- In terms of a group message, once the message is sent, it seems a better approach to clean everything up right away instead of having liggering objects until the async `onDone` is called. 
- Calling the `onDone` async  causes several tests to fail and segmentation faults.
@bzbarsky-apple @mrjerryjohns

Reply:
I broke SendCommandRequest into two functions, one for groups and one for unicast with each their own documentation.